### DISCUSSION TITLE
Problems in connecting Crazyflie Bolt by UART

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1224] [Q&A] [answered]
Hi,
I'm trying to connect the Crazyflie Bolt with an embedded machine vision camera (Jevois). But when I run the scripts in Jevois, it shows the Exception: Could not identify device. The detail information is under below.

1. Goal: Connect and print the log variables of the Crazyflie on the Jevois.

2. Reference: (1) [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/)  in CF
                      (2) [Serial port user guide](http://jevois.org/doc/UserSerial.html) in Jevois

3. Works: (1) Physical Connection: Under the guidance of references, I connected the Bolt with Jevois. 
![jevois](https://github.com/bitcraze/discussions/assets/92587824/f1d21adf-3ac5-48ca-b4f8-167d1fe0a5bd)
![Bolt](https://github.com/bitcraze/discussions/assets/92587824/a52747cd-45ea-41a7-b036-7779a84fc0c2)
    (2) Compiling the crazyflie-firmware with make menuconfig to enbale UART.
    (3) Installing crazyflie-lib-python and pyserial on Jevois
    (4) Prepare example script in [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) with URI = 'serial://dev/ttyS0' on Jevois.

4. Problems: (1) When I run the example script, the console showed error: Exception: Could not identify device.
![console](https://github.com/bitcraze/discussions/assets/92587824/8afdaf03-0b18-477a-a6ba-2ae6d6ff48bf)
(2) And when I wanted to confirm that pyserial indeed finds UART port by "python3 -m serial.tools.list_ports -v". I get "No ports found".
(3) when I run "ls -l /dev/ttyS*", I get"
crw-------    1 root     root      253,   0 Jan  1 00:00 /dev/ttyS0
crw-------    1 root     root      253,   1 Jan  1 00:00 /dev/ttyS1"

BTW, I have tested the Loco Positioning System with Loco Positioning deck installed on this drone. It worked well.

5. Questions: (1) Do you have any suggestions on such problem?
(2) Can CF Bolt connects the Jevois with UART and the Cf client on my computer with radio at the same time?

Looking forward to your reply. Thank you!

**Comment by tobbeanton** [comment]
> Questions: (1) Do you have any suggestions on such problem?

I have newer tried this type of setup but I'm guessing something is missing in python such as pyserial or perhaps not new enough python version?

> Can CF Bolt connects the Jevois with UART and the Cf client on my computer with radio at the same time?

No as it redirects the link over UART instead of the radio. It might be possible to implement code so that two links could be running at the same time but that would be quite a challenging project.

**Comment by Residualstress** [comment]
Thanks for replying!

> I have newer tried this type of setup but I'm guessing something is missing in python such as pyserial or perhaps not new enough python version?

The python version is 3.7.3. And the pyserial version is 3.5.

> No as it redirects the link over UART instead of the radio. It might be possible to implement code so that two links could be running at the same time but that would be quite a challenging project.

So does this mean that when we use AIdeck we can't connect with radio and crazyflie?

In this case, I can't connect to crazyflie via URI because the serial.tools.list_ports function doesn't work properly, but I noticed in crazyflie-lib-python/cflib/cpx/transports.p that the class UARTTransport actually connects directly through device and baudrate. So I want to use this function directly for data transfer. But I can't find any relevant examples, do you have any suggestions?

Note: I have successfully connected my embedded device to an Arduino for data transfer, which proves that its hardware is fine!

**Comment by tobbeanton** [comment]
> So does this mean that when we use AIdeck we can't connect with radio and crazyflie?

The AI-deck doesn't use [CRTP](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/) so it works fine. The problem happens if two hosts would use CRTP at the same time, which is currently not supported.

> In this case, I can't connect to crazyflie via URI because the serial.tools.list_ports function doesn't work properly, but I noticed in crazyflie-lib-python/cflib/cpx/transports.p that the class UARTTransport actually connects directly through device and baudrate. So I want to use this function directly for data transfer. But I can't find any relevant examples, do you have any suggestions?

Perhaps [this](https://www.bitcraze.io/2022/09/using-the-ai-deck-for-crtp-over-wifi-via-cpx/) blog post and its example could be of some help.

Can you update python to a newer version, maybe that is why serial.tools.list_ports function doesn't work properly?

**Reply by Residualstress** [reply]
Thank you, I will try that.

**Comment by Residualstress** [comment, marked as answer]
Hi, finally I found the solution for this problem "couldn't identify devices" in this [topic](https://github.com/pyserial/pyserial/pull/709#issue-1810167929). CF uses pyserial to connect with other services. But pyserial sometimes will ignore the hardware serial port (e.g., /dev/ttyS0) in embedded systems. Pyserial will filter such serial ports in [list_ports_linux.py](https://github.com/pyserial/pyserial/blob/3a6ae5a630589a29af8d8bd45616d7aadb2fb3db/serial/tools/list_ports_linux.py#L103)(Line 103). So [the easiest way](https://github.com/pyserial/pyserial/pull/632/files) is to comment out the line, but make sure there are no new problems first. And the [topic](https://github.com/pyserial/pyserial/pull/709#issue-1810167929) given at the beginning is just as valid.

BTW, this error also could happen when you follow the [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) tutorial with using the mini UART (/dev/ttyS0) to communicate with the Crazyflie.

**Reply by tobbeanton** [reply]
Good find!

> BTW, this error also could happen when you follow the [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) tutorial with using the mini UART (/dev/ttyS0) to communicate with the Crazyflie.

Do you mind elaborating a bit more on this. Do you mean the the tutorial will not work due to your findings?

**Reply by Residualstress** [reply]
Hi, I mean when I follow every step in [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/), the log variables of the Crazyflie could be printed in normal. BTW, I'm using Raspberry Pi4 and Python 3.7.

But when I skip restore /dev/ttyAMA0 as primary UART, i.e., using the mini UART (/dev/ttyS0) to communicate with the Crazyflie (URI = 'serial://ttyS0'), the script will stop at ''with SyncCrazyflie(URI) as scf:''. The error messeage is "Exception: Could not identify device." 

And when I run "python3 -m serial.tools.list_ports -v", The result it returns doesn't include "/dev/ttyS0". So, the error occurs for the following reasons: when you use "with SyncCrazyflie(URI) as scf:" to connect with Crazyflie, It will check if the list returned by "serial.tools.list_ports" contains the URI. (The connection will naturally fail due to our URI is 'serial://ttyS0'.)

The root cause of "ttyS0" not being included in the output of serial.tools.list_ports is in pyserial, and is clearly stated under this  [topic](https://github.com/pyserial/pyserial/issues/489).

So, based on my personal experience, there are two ways to fix this error:

- On Raspberry Pi, you could restore /dev/ttyAMA0 as primary UART to avoid such error.

- For many other embedded systems, their own uart ports are all filtered out by pyserial, and I can only get the program to work by modifying the pyserial, i.e., getting pyserial to correctly find the uart ports in the system. The details of how to do the modification can be found in this [request](https://github.com/pyserial/pyserial/pull/709#issue-1810167929) or the [another one](https://github.com/pyserial/pyserial/pull/632/files).

I hope my experience is useful to you : ).

**Reply by tobbeanton** [reply]
Many thanks for sharing this. I'm sure it will help many others.
